src/libstd/path.rs

```diff
@@ -0,0 +1,2415 @@
+    ///
+    /// The extension is:
+    ///
+    /// * None, if there is no file name;
+    /// * None, if there is no embedded `.`;
+    /// * None, if the file name begins with `.` and has no other `.`s within;
+    /// * Otherwise, the portion of the file name after the final `.`
+    pub fn extension(&self) -> Option<&OsStr> {
+        self.file_name().map(split_file_at_dot).and_then(|(before, after)| before.and(after))
+    }
+
+    /// Creates an owned `PathBuf` with `path` adjoined to `self`.
+    ///
+    /// See `PathBuf::push` for more details on what it means to adjoin a path.
+    pub fn join<P: ?Sized>(&self, path: &P) -> PathBuf where P: AsPath {
+        let mut buf = self.to_path_buf();
+        buf.push(path);
+        buf
+    }
+
+    /// Creates an owned `PathBuf` like `self` but with the given file name.
+    ///
+    /// See `PathBuf::set_file_name` for more details.
+    pub fn with_file_name<S: ?Sized>(&self, file_name: &S) -> PathBuf where S: AsOsStr {
+        let mut buf = self.to_path_buf();
+        buf.set_file_name(file_name);
+        buf
+    }
+
+    /// Creates an owned `PathBuf` like `self` but with the given extension.
+    ///
+    /// See `PathBuf::set_extension` for more details.
+    pub fn with_extension<S: ?Sized>(&self, extension: &S) -> PathBuf where S: AsOsStr {
+        let mut buf = self.to_path_buf();
+        buf.set_extension(extension);
+        buf
+    }
+
+    /// Produce an iterator over the components of the path.
+    pub fn components(&self) -> Components {
+        let prefix = parse_prefix(self.as_os_str());
+        Components {
+            path: self.as_u8_slice(),
+            prefix: prefix,
+            has_physical_root: has_physical_root(self.as_u8_slice(), prefix),
+            front: State::Prefix,
+            back: if has_suffix(self.as_u8_slice(), prefix) { State::Suffix }
+                  else { State::Body },
+        }
+    }
+
+    /// Produce an iterator over the path's components viewed as `OsStr` slices.
+    pub fn iter(&self) -> Iter {
+        Iter { inner: self.components() }
+    }
+}
+
+impl cmp::PartialEq for Path {
+    fn eq(&self, other: &Path) -> bool {
+        iter::order::eq(self.components(), other.components())
+    }
+}
+
+impl PathBuf {
+    fn as_mut_vec(&mut self) -> &mut Vec<u8> {
+        unsafe { mem::transmute(self) }
+    }
+
+    /// Allocate an empty `PathBuf`.
+    pub fn new() -> PathBuf {
+        PathBuf { inner: OsString::from_str("") }
+    }
+
+    /// Allocate a `PathBuf` with contents `path`.
+    pub fn from_path(path: &Path) -> PathBuf {
+        PathBuf { inner: path.as_os_str().to_os_string() }
+    }
+
+    /// Allocate a `PathBuf` with contents `path`.
+    pub fn from_str(path: &str) -> PathBuf {
+        PathBuf { inner: OsString::from_str(path) }
+    }
+
+    /// Allocate a `PathBuf` with contents `path`.
+    pub fn from_os_str(path: &OsStr) -> PathBuf {
+        PathBuf { inner: path.to_os_string() }
+    }
```

Comment:
I mentioned this above as well, but ergonomically we may prefer `fn new<T: AsOsStr>` instead of all the various variants.